src/SettlerSwapper.sol

```diff
@@ -0,0 +1,237 @@
+import {ISettlerBase} from "./interfaces/ISettlerBase.sol";
+import {ISettlerTakerSubmitted} from "./interfaces/ISettlerTakerSubmitted.sol";
+import {IAllowanceHolder} from "./allowanceholder/IAllowanceHolder.sol";
+
+import {IDeployer} from "./deployer/IDeployer.sol";
+
+import {DEPLOYER as DEPLOYER_ADDRESS} from "./deployer/DeployerAddress.sol";
+
+import {Panic} from "./utils/Panic.sol";
+
+library FastDeployer {
+    function fastOwnerOf(IDeployer deployer, uint256 tokenId) internal view returns (address r) {
+        assembly ("memory-safe") {
+            mstore(0x00, 0x6352211e) // selector for `ownerOf(uint256)`
+            mstore(0x20, tokenId)
+
+            if iszero(staticcall(gas(), deployer, 0x1c, 0x24, 0x00, 0x20)) {
+                let ptr := mload(0x40)
+                returndatacopy(ptr, 0x00, returndatasize())
+                revert(ptr, returndatasize())
+            }
+            if or(gt(0x20, returndatasize()), shr(0xa0, r)) { revert(0x00, 0x00) }
+            r := mload(0x00)
+        }
+    }
+
+    function fastPrev(IDeployer deployer, uint128 tokenId) internal view returns (address r) {
+        assembly ("memory-safe") {
+            mstore(0x10, tokenId)
+            mstore(0x00, 0xe2603dc200000000000000000000000000000000) // selector for `prev(uint128)` with `tokenId`'s padding
+
+            if iszero(staticcall(gas(), deployer, 0x0c, 0x24, 0x00, 0x20)) {
+                let ptr := mload(0x40)
+                returndatacopy(ptr, 0x00, returndatasize())
+                revert(ptr, returndatasize())
+            }
+            if or(gt(0x20, returndatasize()), shr(0xa0, r)) { revert(0x00, 0x00) }
+            r := mload(0x00)
+        }
+    }
+}
+
+abstract contract SettlerSwapper {
+    using FastDeployer for IDeployer;
+
+    address internal constant ALLOWANCE_HOLDER_ADDRESS = 0x0000000000001fF3684f28c67538d4D072C22734;
+    IDeployer private constant _DEPLOYER = IDeployer(DEPLOYER_ADDRESS);
+    uint128 private constant _SETTLER_TOKENID = 2;
+
+    error CounterfeitSettler(ISettlerTakerSubmitted counterfeitSettler);
+    error ApproveFailed(IERC20 token);
+
+    function requireValidSettler(ISettlerTakerSubmitted settler) internal view {
+        // Any revert in `ownerOf` or `prev` will be bubbled. Any error in ABIDecoding the result
+        // will result in a revert without a reason string.
+        if (
+            _DEPLOYER.fastOwnerOf(_SETTLER_TOKENID) != address(settler)
+                && _DEPLOYER.fastPrev(_SETTLER_TOKENID) != address(settler)
+        ) {
+            assembly ("memory-safe") {
+                mstore(0x14, settler)
+                mstore(0x00, 0x7a1cd8fa000000000000000000000000) // selector for `CounterfeitSettler(address)` with `settler`'s padding
+                revert(0x10, 0x24)
+            }
+        }
+    }
+
+    modifier validSettler(ISettlerTakerSubmitted settler) {
+        requireValidSettler(settler);
+        _;
+    }
+
+    function _swapAll(
```

Comment:
wondering why do you prefer here to duplicate the entire function instead of just passing computing the amount and reusing the `_swap`. Is it the case that it is a huge improvement?